pkg/util/log/log_test.go

```diff
@@ -499,3 +503,105 @@ func TestStackDepthfLogging(t *testing.T) {
 		})
 	}
 }
+
+func mockScrubBytesWithCount(t *testing.T) *atomic.Int32 {
+	oldScrubber := scrubBytesFunc
+	t.Cleanup(func() { scrubBytesFunc = oldScrubber })
+
+	var counter atomic.Int32
+	counterPtr := &counter
+
+	scrubBytesFunc = func(msg []byte) ([]byte, error) {
+		counterPtr.Add(1)
+		return msg, nil
+	}
+
+	return counterPtr
+}
+
+func testFunctionsScrubbingCount(t *testing.T, funcs []any, args []any) {
+	for _, fun := range funcs {
+		valTy := reflect.TypeOf(fun)
+		require.Equal(t, valTy.Kind(), reflect.Func)
+
+		// get the function name to use as test name
+		val := reflect.ValueOf(fun)
+		fun := runtime.FuncForPC(val.Pointer())
+		require.NotNil(t, fun)
+
+		funcName := fun.Name()
+		if parts := strings.Split(funcName, "/"); len(parts) > 0 {
+			funcName = parts[len(parts)-1]
+		}
+
+		testName := fmt.Sprintf("Scrub Count %s", funcName)
+		t.Run(testName, func(t *testing.T) {
+			// create a slice of reflect.Value from the args
+			reflArgs := make([]reflect.Value, 0, len(args))
+			for _, arg := range args {
+				reflArgs = append(reflArgs, reflect.ValueOf(arg))
+			}
+
+			counter := mockScrubBytesWithCount(t)
+			val.Call(reflArgs)
+			require.Equal(t, 1, int(counter.Load()))
+		})
+	}
+}
+
+func TestLoggerScrubbingCount(t *testing.T) {
+	var b bytes.Buffer
+	w := bufio.NewWriter(&b)
+	l, err := seelog.LoggerFromWriterWithMinLevelAndFormat(w, seelog.TraceLvl, "[%LEVEL] %FuncShort: %Msg")
+	require.NoError(t, err)
+	SetupLogger(l, "trace")
+
+	// package public methods
+	testFunctionsScrubbingCount(
+		t,
+		[]any{Trace, Debug, Info, Warn, Error, Critical},
+		[]any{"a", "b"},
+	)
+	testFunctionsScrubbingCount(
+		t,
+		[]any{Tracef, Debugf, Infof, Warnf, Errorf, Criticalf},
+		[]any{"a %s c", "b"},
+	)
+	testFunctionsScrubbingCount(
+		t,
+		[]any{TracefStackDepth, DebugfStackDepth, InfofStackDepth, WarnfStackDepth, ErrorfStackDepth, CriticalfStackDepth},
+		[]any{1, "a %s c", "b"},
+	)
+	testFunctionsScrubbingCount(
+		t,
+		[]any{Tracec, Debugc, Infoc, Warnc, Errorc, Criticalc},
+		[]any{"a b", "1 %s 3", "2"},
+	)
+	testFunctionsScrubbingCount(
+		t,
+		[]any{TracecStackDepth, DebugcStackDepth, InfocStackDepth, WarncStackDepth, ErrorcStackDepth, CriticalcStackDepth},
+		[]any{"a b", 1, "1 %s 3", "2"},
+	)
+	testFunctionsScrubbingCount(
+		t,
+		[]any{TraceFunc, DebugFunc, InfoFunc, WarnFunc, ErrorFunc, CriticalFunc},
+		[]any{func() string { return "a b" }},
+	)
+
+	// loggerPointer methods
+	testFunctionsScrubbingCount(
+		t,
+		[]any{logger.trace, logger.debug, logger.info, logger.warn, logger.error, logger.critical},
+		[]any{"a b"},
+	)
+	testFunctionsScrubbingCount(
+		t,
+		[]any{logger.tracef, logger.debugf, logger.infof, logger.warnf, logger.errorf, logger.criticalf},
+		[]any{"a %s c", "b"},
+	)
+	testFunctionsScrubbingCount(
+		t,
+		[]any{logger.traceStackDepth, logger.debugStackDepth, logger.infoStackDepth, logger.warnStackDepth, logger.errorStackDepth, logger.criticalStackDepth},
+		[]any{"a b", 1},
+	)
```

Comment:
Using a custom pattern for running multiple tests within a single test function, while fun, might make it more difficult in the future to have a consistent testing codebase pattern. 

By grepping the code base, we use the [Table Testing Pattern](https://golang.cafe/blog/golang-table-test-example). You can run `git grep 'range tests'` the output.

Positive things about using table testing:
- No need to use reflection to get the test name 
```go
// get the function name to use as test name
val := reflect.ValueOf(fun)
fun := runtime.FuncForPC(val.Pointer())
require.NotNil(t, fun)
```

- Easier to read the different test cases as the test body is on the `TestLoggerScrubbingCount` function

What do you think?

Reply:
I considered it initially but considering there are 6 "kinds" of functions taking different arguments, with 6 functions for each, I wanted to avoid having to write 36 structs with name + function + args 😄

Reply:
I agree that `reflect` makes code less readable but if the alternative is several hundreds lines of code I still think it's better

Reply:
I see what you mean. 

Maybe I wasn't clear with what I suggested, here is a code sample:

```go
func TestLoggerScrubbingCount(t *testing.T) {
	var b bytes.Buffer
	w := bufio.NewWriter(&b)
	l, err := seelog.LoggerFromWriterWithMinLevelAndFormat(w, seelog.TraceLvl, "[%LEVEL] %FuncShort: %Msg")
	require.NoError(t, err)
	SetupLogger(l, "trace")

	tests := []struct {
		name  string
		funcs []any
		args  []any
	}{
		{"Public methods", []any{Trace, Debug, Info, Warn, Error, Critical}, []any{"a", "b"}},
		{"Public methods with format options", []any{Tracef, Debugf, Infof, Warnf, Errorf, Criticalf}, []any{"a %s c", "b"}},
	}

	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			for _, fun := range test.funcs {

				val := reflect.ValueOf(fun)

				// create a slice of reflect.Value from the args
				reflArgs := make([]reflect.Value, 0, len(test.args))
				for _, arg := range test.args {
					reflArgs = append(reflArgs, reflect.ValueOf(arg))
				}

				counter := mockScrubBytesWithCount(t)
				val.Call(reflArgs)
				require.Equal(t, 1, int(counter.Load()))
			}
		})
	}
}
``` 

I was planning to group the function into a single test name. 

What do you think?

Reply:
@GustavoCaso I refactored the test to have one subtest per kind of function, which simplifies the code, and used `assert` to avoid failing the whole subtest in case of error, that way we still have individual errors per function
[`5a9d83c` (#22013)](https://github.com/DataDog/datadog-agent/pull/22013/commits/5a9d83cb1e4dc6e41f8269fc2de4fba9680627bf)

Reply:
Thank you so much for making those changes, @pgimalac. The test is way more readable now 🎉